### ISSUE TITLE
databaseid 

### ISSUE THREAD

**Issue by SmoxTV** [TheHolyWaffle/TeamSpeak-3-Java-API#134]
Hey,

i need your help.

First, sorry for my English i´m from Germany.
How can I get the Clientdatabaseid in the Commands ?

Thanks for help <3

**Comment by Firedroide** [comment]
That really depends on what information is available to you 😛 

You can either call [`TS3Api#getClients`](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/src/main/java/com/github/theholywaffle/teamspeak3/TS3Api.java#L2090), which gets you a list of clients from which you need to filter out your client, or call [`TS3Api#getClientInfo`](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/src/main/java/com/github/theholywaffle/teamspeak3/TS3Api.java#L2048) which gets you a `ClientInfo` directly. There's also [`TS3Api#getClientByUId`](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/src/main/java/com/github/theholywaffle/teamspeak3/TS3Api.java#L2028) if you only know the client's unique ID but not their client ID.

On either one of those client objects you can call [`.getDatabaseId()`](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/src/main/java/com/github/theholywaffle/teamspeak3/api/wrapper/Client.java#L86) to get that client's database ID.

I hope I was able to help you :smile:

**Comment by SmoxTV** [comment]
Have you skype :p

Name: SmoxWeb.xyz

**Comment by Firedroide** [comment]
No, but you can always contact me on [Gitter](https://gitter.im/Firedroide) if you want to chat.

**Comment by SmoxTV** [comment]
IT doesn´t work :( :-1:   (JAVA)

**Comment by Firedroide** [comment]
That literally gives me nothing to work with 😫 

At least tell me *what* doesn't work or if there is an error message.

And you could be extra awesome by posting the two things requires for a good issue report:
- Your code
- The console log

And if you do, please use some code tags, like this:

    ```
    your code here
    ```

**Comment by SmoxTV** [comment]
`package listener;

import java.util.Map;

import com.github.theholywaffle.teamspeak3.TS3Api;
import com.github.theholywaffle.teamspeak3.TS3Query;
import com.github.theholywaffle.teamspeak3.api.ClientProperty;
import com.github.theholywaffle.teamspeak3.api.TextMessageTargetMode;
import com.github.theholywaffle.teamspeak3.api.event.*;
import com.github.theholywaffle.teamspeak3.commands.CChannelClientAddPerm;

public class CMD_Verify extends TS3EventAdapter implements TS3Event 
{
	public void Client(Map<String, String> map) {
	}
	public int getDatabaseId() {
		return getInt(ClientProperty.CLIENT_DATABASE_ID);
	}
	private int getInt(ClientProperty clientDatabaseId) {
		return 0;
	}
	TS3Api api;
	public CMD_Verify(final TS3Api api){
		this.api = api;
		final int clientId = api.whoAmI().getId();
		api.registerEvent(TS3EventType.TEXT_CHANNEL, 0);
		api.addTS3Listeners(new TS3EventAdapter() {
			public void onTextMessage(TextMessageEvent e) {
				final int userid = e.getInvokerId();
				final int userdbid = api.whoAmI().getDatabaseId();
			if (e.getTargetMode() == TextMessageTargetMode.CLIENT && e.getInvokerId() != clientId) {
			String message = e.getMessage().toLowerCase();
                if (message.equalsIgnoreCase("!Verify")) {
                    api.sendPrivateMessage(userid, "Akzeptierst du die AGB´s? [b]!Verify JA | !Verify NEIN[/b]");
			}
                if (message.equalsIgnoreCase("!Verify NEIN")) {
                    api.kickClientFromServer("Verify | Du hast NEIN ausgewählt!", userid);
			}
            	if (message.equalsIgnoreCase("!Verify JA")) {				
            		api.addClientToServerGroup(7, ClientProperty.CLIENT_DATABASE_ID);				
            		api.kickClientFromServer("Verify | Bitte joine neu!", userid);
            	}
			}
		}
		});
		}
	@Override
	public void fire(TS3Listener arg0) {
		// TODO Auto-generated method stub
		
	}
		}

`

**Comment by SmoxTV** [comment]
can you help me?

**Comment by Firedroide** [comment]
Writing this stuff takes some time, man 😛 Have some patience, please.

```
if (message.equalsIgnoreCase("!verify ja")) {
    ClientInfo info = api.getClientInfo(userid);
    if (info == null) return; // Client quit immediately after this message or command failed.
    api.addClientToServerGroup(7, info.getDatabaseId());
//    api.kickClientFromServer("Verify | Bitte joine neu!", userid); <-- What on earth are you doing this for??
}
```

You should've just told me that you didn't understand my first message. I should've explained it in simpler terms in the first place. Sorry for that.

Some other advice:

- `CMD_Verify` does not need to extend `TS3EventAdapter` or implement `TS3Event`. Remove those statements (including `public void fire(TS3Listener arg0)`).
- Remove the methods `getDatabaseId()`, `getInt(ClientProperty clientDatabaseId)` and the **field** `TS3Api api`. Copying stuff out from decompiled classes you don't fully understand is almost always a bad idea, by the way.
- Consider reading some of the [examples](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/tree/master/example).
- Consider using some `else if`s or `switch` blocks.

**Comment by SmoxTV** [comment]
Thanks its works and: 
Can I retrieve e.g. when the player has member he gets another welcome message

**Comment by Firedroide** [comment]
> when the player has member

Sorry, I have no idea what you mean by "the player has a member".
The German version you sent me on Gitter doesn't make more sense, either.

**Comment by Firedroide** [comment]
Oh, I get it. You mean "when a player is a member of a certain server group".

Sure. Register for `TS3EventType.SERVER` and add a new `TS3EventAdapter` event listener with

```java
@Override
public void onClientJoin(ClientJoinEvent e) {
    Client client = new Client(e.getMap()); // This is a bit hacky. It'll continue working, but there will be a better way for it in the future
    if (client.isInServerGroup(yourServerGroupIdHere)) {
        // Send them a special welcome message
    } else {
        // Send them a default welcome message
    }
}
```

**Comment by Kakifrucht** [comment]
I think he means he wants to check if a user already accepted the terms and send a different message then.
Just use the API to check the server group he is in, if he is in group with id 7 send this message, else send verify.

**Comment by Firedroide** [comment]
Sniped! 😛

**Comment by SmoxTV** [comment]
Omg thanks @Firedroide  :) and sorryyyyyyyyyyyyyy for my english :)

**Comment by SmoxTV** [comment]
one problem the message dont work
`
package listener;




import com.github.theholywaffle.teamspeak3.TS3Api;
import com.github.theholywaffle.teamspeak3.api.event.ClientJoinEvent;
import com.github.theholywaffle.teamspeak3.api.event.TS3EventAdapter;
import com.github.theholywaffle.teamspeak3.api.event.TextMessageEvent;
import com.github.theholywaffle.teamspeak3.api.wrapper.*;
import com.github.theholywaffle.teamspeak3.api.wrapper.ClientInfo;
import com.github.theholywaffle.teamspeak3.api.wrapper.Message;


public class Willkommen extends TS3EventAdapter{
	TS3Api api;
	public Willkommen(TS3Api api){
		this.api = api;
		api.registerAllEvents();
		api.addTS3Listeners(new TS3EventAdapter() {
	public void onClientJoin (ClientJoinEvent e) {
	   final int userid = e.getInvokerId();
		Client client = new Client(e.getMap());
		if (client.isInServerGroup(7)) {
	        api.sendPrivateMessage(userid, "Willkommen :)");
	    } else {
	        api.sendPrivateMessage(userid, "Willkommen.");
	    }
	}
	});
	}
}
`

**Comment by Firedroide** [comment]
Yes, that's because `ClientJoinEvent` does not have an invoker. An invoker is an external client who caused this event to happen. Now, I'm sure you'll agree that there's no way another client can cause someone to connect to a server, right? Therefore, there's no invoker ID.

And you would've known that this was the case if you had looked at the console output of your program, specifically at the line that was prefixed with `[SEVERE]` and the two lines just above that message.

Use the client ID from `client.getId()` instead.

**Comment by SmoxTV** [comment]
Okay :D

**Comment by SmoxTV** [comment]
It works :+1:  nice you are good :D

**Comment by SmoxTV** [comment]
Can you help me with this ;DDDD

`package listener;

import java.util.Map;

import com.github.theholywaffle.teamspeak3.TS3Api;
import com.github.theholywaffle.teamspeak3.TS3Query;
import com.github.theholywaffle.teamspeak3.api.ClientProperty;
import com.github.theholywaffle.teamspeak3.api.TextMessageTargetMode;
import com.github.theholywaffle.teamspeak3.api.event.*;
import com.github.theholywaffle.teamspeak3.api.wrapper.Client;
import com.github.theholywaffle.teamspeak3.api.wrapper.ClientInfo;
import com.github.theholywaffle.teamspeak3.commands.CChannelClientAddPerm;

public class CMD_Ruhe extends TS3EventAdapter implements TS3Event 
{
    public void Client(Map<String, String> map) {
    }
    public int getDatabaseId() {
        return getInt(ClientProperty.CLIENT_DATABASE_ID);
    }
    private int getInt(ClientProperty clientDatabaseId) {
        return 0;
    }
    TS3Api api;
    public CMD_Ruhe(final TS3Api api){
        this.api = api;
        final int clientId = api.whoAmI().getId();
        api.registerEvent(TS3EventType.TEXT_CHANNEL, 0);
        api.addTS3Listeners(new TS3EventAdapter() {
            public void onTextMessage(TextMessageEvent e) {
            if (e.getTargetMode() == TextMessageTargetMode.CLIENT && e.getInvokerId() != clientId) {
            String message = e.getMessage().toLowerCase();
                if (message.equalsIgnoreCase("ruhe")) {
            		Client client = new Client(e.getMap()); 
            	    if (client.isInServerGroup(7)) {
                      api.sendPrivateMessage(client.getId(), "Du wurdest nun ihn der Gruppe Ruhe eingeteilt.");
            	    } else {
            	    	api.sendPrivateMessage(client.getId(), "Wie es wohl ausieht hast du für diesen Befehl keine Rechte.");
            	    }
                }
            }
        }
        });
        }
    @Override
    public void fire(TS3Listener arg0) {
    }
        }
`

**Comment by SmoxTV** [comment]
you are here ? @Firedroide

**Comment by SmoxTV** [comment]
and @Firedroide 

1. Can I block certain countrys with Query Access?
2. Can certain groups, certain commands executing? 
3. How can I trace the ISP from a client?

(MY ENGLISH = BULLSHIT)

**Comment by Firedroide** [comment]
Please, figure it out yourself. That's the only way you'll actually learn something from all this and not rely on my help to get anything done.

I'm not here to do the programming for you. I am also not a search engine.
I've already answered your questions three times and I've provided you with all the tools needed for you to be able to program a clever server query client.

If you need more help, use our documentation, the JavaDoc, the examples, the source code, a clever IDE and maybe even a good search engine.

**Comment by SmoxTV** [comment]
@Firedroide WHY ?

**Comment by SmoxTV** [comment]
how can I open this Java document?

**Comment by Kakifrucht** [comment]
You obviously lack a lot of very basic programming knowledge. As stated, how about googling what a "JavaDoc" is and learning a bit more about Java and programming in general. It's all there.

**Comment by SquadScope** [comment]
@SmoxTV I can send you a part of my Code, if you want to, I've done some commands like "!register"... And new users will recieve some pokes about the server rules...

P.S.: I'm from germany too! :D

**Comment by SmoxTV** [comment]
@SquadScope  Okay sorry for the late anwser add me on skype : smoxweb.xyz or come teamspeak: swradio.wtf
